lib/xmain/opts.go

```diff
@@ -0,0 +1,132 @@
+package xmain
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"oss.terrastruct.com/cmdlog"
+	"oss.terrastruct.com/xos"
+)
+
+type Opts struct {
+	args  []string
+	flags *pflag.FlagSet
+	env   *xos.Env
+	log   *cmdlog.Logger
+
+	registeredEnvs []string
+}
+
+func NewOpts(env *xos.Env, args []string, log *cmdlog.Logger) *Opts {
+	flags := pflag.NewFlagSet("", pflag.ContinueOnError)
+	flags.SortFlags = false
+	flags.Usage = func() {}
+	flags.SetOutput(io.Discard)
+	return &Opts{
+		args:  args,
+		flags: flags,
+		env:   env,
+		log:   log,
+	}
+}
+
+func (o *Opts) Help() string {
+	b := &strings.Builder{}
+	o.flags.SetOutput(b)
+	o.flags.PrintDefaults()
+
+	if len(o.registeredEnvs) > 0 {
+		b.WriteString("\nYou may persistently set the following as environment variables (flags take precedent):\n")
+		for i, e := range o.registeredEnvs {
+			s := fmt.Sprintf("- $%s", e)
+			if i != len(o.registeredEnvs)-1 {
+				s += "\n"
+			}
+			b.WriteString(s)
+		}
+	}
+
+	return b.String()
+}
+
+func (o *Opts) Int64(envKey, flag, shortFlag string, defaultVal int64, usage string) *int64 {
+	if envKey != "" {
+		if o.env.Getenv(envKey) != "" {
+			envVal, err := strconv.ParseInt(o.env.Getenv(envKey), 10, 64)
+			if err != nil {
+				o.log.Error.Printf(`ignoring invalid environment variable %s. Expected int64. Found "%v".`, envKey, envVal)
+			} else if envVal != defaultVal {
+				defaultVal = envVal
+			}
+		}
+		o.registeredEnvs = append(o.registeredEnvs, envKey)
+	}
+
+	return o.flags.Int64P(flag, shortFlag, defaultVal, usage)
+}
+
+func (o *Opts) String(envKey, flag, shortFlag string, defaultVal, usage string) *string {
+	if envKey != "" {
+		if o.env.Getenv(envKey) != "" {
+			envVal := o.env.Getenv(envKey)
+			if envVal != defaultVal {
+				defaultVal = envVal
+			}
+		}
+		o.registeredEnvs = append(o.registeredEnvs, envKey)
+	}
+
+	return o.flags.StringP(flag, shortFlag, defaultVal, usage)
+}
+
+func (o *Opts) Bool(envKey, flag, shortFlag string, defaultVal bool, usage string) *bool {
+	if envKey != "" {
+		if o.env.Getenv(envKey) != "" {
+			envVal := o.env.Getenv(envKey)
+			if !boolyEnv(envVal) {
+				o.log.Error.Printf(`ignoring invalid environment variable %s. Expected bool. Found "%s".`, envKey, envVal)
+			} else if (defaultVal && falseyEnv(envVal)) ||
+				(!defaultVal && truthyEnv(envVal)) {
+				defaultVal = !defaultVal
+			}
+		}
+		o.registeredEnvs = append(o.registeredEnvs, envKey)
+	}
+
+	return o.flags.BoolP(flag, shortFlag, defaultVal, usage)
+}
+
+func boolyEnv(s string) bool {
+	return falseyEnv(s) || truthyEnv(s)
+}
+
+func falseyEnv(s string) bool {
+	return s == "0" || s == "false" || s == "f"
```

Comment:
looking at this again, i'm also inclined to add "yes/no". 

`f`/`t` is a stretch, ur right. these aren't what we recommend tbc, just things users might instinctively set

Reply:
No absolutely not. true or false only. google "yaml no"

Reply:
No, such instinctive inputs may be instinctive but that doesn't mean we should accept them. I have no argument here other than I don't want more ways to do things than necessary and true, false 1,0 is already a compromise on my end.

Reply:
feels overly dogmatic to me. the average user isn't running on archlinux and set 20 million env vars before. i suppose this comes down to me thinking good programs are dummy-proof and your perspective that good programs follow conventions and don't perpetuate bad ones. 

but nbd, tiny hill, i'll defer to ur stronger opinion and keep it as true/false 1/0.

Reply:
> give a clear error message and have them fix it? Is that not uniliterally superior?

well they still have to re-set the var when we knew what they meant, so not unilaterally. if d2lang.com is the one true domain, would you show an error for d2-lang.com saying "please go to d2lang.com" or just redirect?

but i did think of something else that supports not adding yes/no, and that's that we'd have to support it everywhere in the future. install flags, future cli flags, env vars, idk a few might even expect them to be valid inputs for booleans in d2. so, that'd be annoying and not worth it.

Reply:
> well they still have to re-set the var when we knew what they meant, so not unilaterally. if d2lang.com is the one true domain, would you show an error for d2-lang.com saying "please go to d2lang.com" or just redirect?

The difference there is that there is only one d2lang.com and so it doesn't spread like a virus the way it does with configuration parameters.

We could have tens of variables. And now you end up with half your config using one and the other half using another and you can't remember what accepts what. And other people using other inputs that you don't recognize and aren't sure if they would work. Having more ways to do things is not a good idea when you can nudge people easily in the right direction. It's only when you have no way to nudge or document should you be overly permissive in the inputs you accept. Which is frankly never.

Let's say we have a norwegian trying to set a variable that represents his country that he wants input into a D2 diagram and he uses `no`. Now he accidentally sets that on the watch env variable instead. Because we didn't produce a clear error, he slapped his head around on his laptop for a few hours until he realizes his mistake. We could have so easily saved him his time, so easily.

And then should case matter? Is `TRUE` and `FALSE` acceptable?

YAML already ran the experiment and we know the results, it's not a good idea. Seriously, google around. Especially `no` and the Norwegians, it's the classic example of YAML being braindead. There's a reason you don't see it in any good programming language. There's immense value in keeping things predictible.

Reply:
Another example, should `layouts` be an accepted subcommand or should it error and recommend that you might have meant `d2 layout`?

Reply:
Circling back to the previous PR, are we set that empty is `false` and anything is `true`?

Reply:
Though what did you mean by get around it with `.env`? What is `.env`?

Reply:
> Though what did you mean by get around it with .env? What is .env?

It's quite common, specially with docker, that you have `.env` local file with all your envs that are used to configure the current app. IDEs can load it too, making configs easier in these cases
https://www.npmjs.com/package/dotenv